KAFKA-9138: Add system test for relational joins

vvcephei opened the pull request:
Adds a system test to verify the new foreign-key join introduced in KIP-213.

Also, adds a unit test and integration test to verify the test logic itself.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

vvcephei committed:
KAFKA-9138: Add system test for relational joins

vvcephei commented on bin/kafka-run-class.sh:
```diff
@@ -74,11 +74,11 @@ done
   clients_lib_dir=$(dirname $0)/../clients/build/libs
   streams_lib_dir=$(dirname $0)/../streams/build/libs
-  rocksdb_lib_dir=$(dirname $0)/../streams/build/dependant-libs-${SCALA_VERSION}
+  streams_dependant_clients_lib_dir=$(dirname $0)/../streams/build/dependant-libs-${SCALA_VERSION}
```
The system test code uses Hamcrest matchers to make its assertions, so I needed to add the hamcrest jars to the classpath as well. It seemed like it would be less confusing to name this variable what it is... the directory containing Streams's dependencies, not just RocksDB.

vvcephei commented on build.gradle:
```diff
@@ -1211,6 +1211,7 @@ project(':streams') {
     from (configurations.testRuntime) {
       include('slf4j-log4j12*')
       include('log4j*jar')
+      include('*hamcrest*')
```
copying over hamcrest for the system test.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/RelationalSmokeTestIntegrationTest.java:
```diff
@@ -0,0 +1,194 @@
+import java.util.UUID;
+
+@Category(IntegrationTest.class)
+public class RelationalSmokeTestIntegrationTest {
```
Pretty straightforward... an integration test for the smoke test for the relational joins system test.

I've found it useful to have integration tests covering some of the system test logic:
* it's hard to debug the system tests themselves
* it's also kind of hard to know if the system tests are really doing anything, or just giving a false sense of security (I have found this to be the case for the smoke test in the past)

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/integration/RelationalSmokeTestIntegrationTest.java:
```diff
@@ -0,0 +1,194 @@
+            driver.exception().printStackTrace();
+            throw new AssertionError(driver.exception());
+        }
+        Assert.assertTrue(driver.passed());
```
If it fails, it'll log the reason why.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,954 @@
+        return Arrays.copyOf(TOPICS, TOPICS.length);
+    }
+
+    public static class Article {
```
This test builds on a basic relational data caricature: a set of articles, each with a set of comments on them.

We're doing a couple of "relational data" operations to derive the "augmented" equivalents of the input data:
* traverse the `Comment-[articleId]->Article` relationship to extract the prefix of the article text and add it to the comment's data structure.
* aggregate the number of comments on each article and add it to the article's data structure.

The first operation is using the new feature. The second was already supported in Streams. I'm thinking that if this test proves to be stable, we can flesh it out even more to eventually subsume the other smoke test.

guozhangwang replied:
> I'm thinking that if this test proves to be stable, we can flesh it out even more to eventually subsume the other smoke test.

Totally, we have too many overlapping `XXXSmokeTest` now and I think we should just consider having one and also trim out system test suite too.

Could you file a JIRA as a tech debt cleanup?

guozhangwang replied:
Also could you actually add a javadoc section on top of the class with the above description about the application's logic? 

#`put your PR comment into code comment` :P

vvcephei replied:
Filed https://issues.apache.org/jira/browse/KAFKA-9289

vvcephei commented on tests/kafkatest/tests/streams/templates/log4j_template.properties:
```diff
@@ -0,0 +1,31 @@
+# Licensed to the Apache Software Foundation (ASF) under one or more
```
There's no way to explicitly include a template by path, it just has to be at this specific relative path to the Service class.

guozhangwang replied:
As long as only one file needed per suite (like `streams/template`) I think it's fine.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,959 @@
+                    mkEntry(StreamsConfig.BOOTSTRAP_SERVERS_CONFIG, broker),
+                    mkEntry(StreamsConfig.APPLICATION_ID_CONFIG, application),
+                    mkEntry(StreamsConfig.CLIENT_ID_CONFIG, id),
+                    mkEntry(StreamsConfig.PROCESSING_GUARANTEE_CONFIG, StreamsConfig.EXACTLY_ONCE),
```
Note, running with EOS so that we can verify the results are exactly correct.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,959 @@
+            return kafkaStreams;
+        }
+
+        public static boolean verifySync(final String broker, final Instant deadline) throws InterruptedException {
```
Note for those familiar with the other smoke test:

This verification works by consuming the same inputs as the topology and then computing what the topology should have computed, given those inputs. Thus, it doesn't depend on every message from the producers actually being delivered exactly once (which is responsible for quite a bit of complexity and flakiness in the other test).

guozhangwang replied:
That's great!

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,959 @@
+                boolean printedConsumedAugmentedComment = false;
+                boolean passed = false;
+
+                while (!passed && Instant.now().isBefore(deadline)) {
```
This also builds on another lesson learned from the other test. Instead of trying to figure out when we've consumed "all" the data, we just keep trying to get a passing result until the deadline. The verification logic itself checks to see if we have gotten at least some input data, so it won't pass trivially.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,959 @@
+                        printedConsumedAugmentedArticle = true;
+                    }
+                    if (!printedConsumedAugmentedComment && !consumedAugmentedComments.isEmpty()) {
+                        LOG.info("Consumed first AugmentedComment");
```
These messages are used by the python code to wait for until the first result output has been observed. This is probably a better wait condition than what we used in the other smoke test (a line printed in a peek in the topology itself), because it means not only that a record got polled, but that it got fully processed and committed in a transaction to the output topic.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,959 @@
+            }
+        }
+
+        static boolean verifySync(final boolean logResults,
```
Here's the actual input vs. output data integrity verification.

vvcephei commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTestTest.java:
```diff
@@ -0,0 +1,105 @@
+import static org.hamcrest.CoreMatchers.startsWith;
+import static org.hamcrest.MatcherAssert.assertThat;
+
+public class RelationalSmokeTestTest extends SmokeTestUtil {
```
This is the unit test for the smoke test logic. Its purpose is just to quickly catch any bugs we might introduce in the smoke test in the future.

vvcephei commented on tests/kafkatest/tests/streams/streams_relational_smoke_test.py:
```diff
@@ -0,0 +1,124 @@
+from kafkatest.tests.kafka_test import KafkaTest
+
+
+class StreamsRelationalSmokeTestService(StreamsTestBaseService):
```
Trying something new here... Rather than defining the service for this test in a separate file together with a bunch of unrelated services for other tests, I'm proposing to define it right here in the same file as the test itself. That way, you can see all the relevant logic for this one test in one place.

The only bummer I found was that the log4j config template is required to be at a specific relative path to the service class, so I had to copy it over (it's part of this PR as well). Of course, if we like this pattern, and want to migrate other tests to it, then the template will already be in place.

vvcephei commented on tests/kafkatest/tests/streams/streams_relational_smoke_test.py:
```diff
@@ -0,0 +1,123 @@
+
+    @cluster(num_nodes=8)
+    @matrix(crash=[False, True])
+    def test_streams(self, crash):
```
One thing to note is the avoidance of log_monitor, which doesn't work the way that it seems like we thought it works. Instead, we're structuring our ssh/grep commands to make the minimum ordering assumptions and be correct by looking for unique events.

guozhangwang replied:
Could you elaborate a bit more about `log_monitor` why it does not work as expected?

vvcephei replied:
The log monitor looks up the current rail position of the file just once when you create it. Subsequently, each grep invocation uses the exact same starting point in the file. 

I’ve noticed that some of our tests seem to assume that the monitor maintains a cursor in the file, but it does not. 

The one nice thing about the monitor is that within it, you’re guaranteed not to see any log messages from before the time you created it, but in practice there are still too many opportunities to miss something or make an incorrect assumption in the test. 

I think a more robust approach is to just expect to search the entire log each time and figure out how to phrase the query as an invariant.

vvcephei reviewed (commented):
Adding a new smoke test for "relational data processing" use cases. The existing smoke test is more for "stream processing" use cases. I know this is somewhat a false dichotomy, but what I mean is that we needed a more traditional RDB-style data set for this logic.

I also took the opportunity to try setting up the system test a little differently, aiming to incorporate several lessons learned from the flakiness of other tests.

vvcephei commented:
Also, I've started a few system test runs:

the new test with `--repeat=90`:
Passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-07--001.1573195705--vvcephei--smoke-test-fk--adb4237/report.html

all streams tests:
* http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-08--001.1573197173--vvcephei--smoke-test-fk--adb4237/report.html (unrelated failure)
* http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-08--001.1573199816--vvcephei--smoke-test-fk--c331c04/report.html (passed)
* http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-11-08--001.1573199846--vvcephei--smoke-test-fk--c331c04/report.html (passed)

vvcephei commented:
Note: the new system test fails some checkstyle complexity constraints. I think it's partially caused by the amount of logic in one file. I considered breaking it up into several files in a package, but I'm on the fence. The alternative is just to exclude this class from those checkstyle rules (which I've done in this PR)... WDYT?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/RelationalSmokeTestIntegrationTest.java:
```diff
@@ -0,0 +1,187 @@
+            // wait for verification to finish
+            driver.join(IntegrationTestUtils.DEFAULT_TIMEOUT);
+        } finally {
+            // whether or not the assertions failed, tell all the streams instances to stop
```
Could we do the same pattern as I did fixing some flakiness of the SmokeTest integration test? Just wait for the state transition in listener. Found sometimes it could cause many more concurrent instances running.

vvcephei replied:
With a little distance, I think maybe we don’t need this integration test after all. It was useful in development, but I think I’ll just drop it. We’ll still have the unit test for the smoke test logic, but otherwise we’ll just rely on the system test itself. 

If we have trouble with the system test later on, we can always reintroduce this integration test at that point.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/integration/RelationalSmokeTestIntegrationTest.java:
```diff
@@ -0,0 +1,187 @@
+
+            // let the oldest client die of "natural causes"
+            if (clients.size() >= 3) {
+                clients.remove(0).close(Duration.ZERO);
```
Here could we wait for the client to complete close via the state listener?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,962 @@
+                '}';
+        }
+
+        public static class CommentSerializer implements Serializer<Comment> {
```
Ahh now I remember why I still love Scala (not a comment :P).

vvcephei replied:
For real.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,962 @@
+
+        public static DataSet generate(final int numArticles, final int numComments) {
+            // generate seven days' worth of data, starting right now (to avoid broker retention/compaction)
+            final int timeSpan = 1000 * 60 * 60 * 24 * 7;
```
I'd suggest we put the time span as 4 days to be on the safer side, the 7 day log retention may kicks in a bit earlier.

vvcephei replied:
Just to make sure, this is seven days in the future we’re talking about. Does log retention use “stream time”? I thought it was always wall clock. 

I’ll change it anyway. 7 days was arbitrary.

guozhangwang replied:
It is wall clock indeed.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,962 @@
+
+
+            final KTable<Integer, Long> commentCounts =
+                comments.groupBy((key, value) -> new KeyValue<>(value.getArticleId(), (short) 1),
```
Nice one on projecting the value!

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,962 @@
+                                 Grouped.with(Serdes.Integer(), Serdes.Short()))
+                        .count();
+
+            final ValueJoiner<Article, Long, AugmentedArticle> fkJoiner =
```
Why not putting this joiner as the AugmentedArticle static as well?

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,962 @@
+            final Map<Integer, Long> commentCounts = new TreeMap<>();
+
+            for (final RelationalSmokeTest.AugmentedComment augmentedComment : consumedAugmentedComments.values()) {
+                assertThat(pass, report, "comment missing", consumedComments, hasKey(augmentedComment.getKey()));
```
Could we include specific article / comment information in the report as well like the line 894 below?

Also I'm wondering if multiple of the assertions have failed, would the later report replace the older one?

vvcephei replied:
Will do. No, they are just all concatenated in the report.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/tests/RelationalSmokeTest.java:
```diff
@@ -0,0 +1,962 @@
+                        consumedComments.size()
+                    );
+                } else {
+                    LOG.error("Evaluation failed\n{}", report);
```
When error, could we print all the received records from all four topics too? Just 2 cents from my past flaky test fighting experience.

guozhangwang commented on tests/kafkatest/tests/streams/streams_relational_smoke_test.py:
```diff
@@ -0,0 +1,123 @@
+
+        driver.node.account.ssh("grep 'Smoke test complete: passed' %s" % driver.LOG_FILE, allow_fail=False)
+
+        driver.stop()
```
Should we close processor3 at the end?

vvcephei replied:
Yes. Good call.

guozhangwang commented on tests/kafkatest/tests/streams/streams_relational_smoke_test.py:
```diff
@@ -0,0 +1,123 @@
+
+        processor1.stop_nodes(not crash)
+
+        processor3 = StreamsRelationalSmokeTestService(self.test_context, self.kafka, "application", "processor3")
```
In other tests we start all three processors, and the rolling bounce them; here we start 2, shutdown 1, and start a new one, and shutdown another. Is it intentional and what's the motivation?

vvcephei replied:
I wanted to make sure that the computation survives state shuffling as a result of processor 3 having to recover all of processor 1's state, and then subsequently all of processor 2's state.

Conversely, if we just do a rolling bounce, then instances would still be able to make use of the state from before.

Do you think this makes sense, or is it just fuzzy thinking on my part?

guozhangwang replied:
Okay I think this makes sense, let's just follow this pattern then.

guozhangwang reviewed (commented):
Made a pass.

vvcephei committed:
code review

vvcephei commented:
Ok, @guozhangwang , I have _finally_ finished addressing your comments. Can you take another look?

I'm on a flaky network right now, so I can't run the system tests. If you have a chance, and you feel it's necessary, maybe you can kick them off again. Otherwise, I'll do it tonight.

Thanks for your review, by the way!

guozhangwang commented:
The new commit lgtm, I've triggered https://jenkins.confluent.io/job/system-test-kafka-branch-builder/3629/. Please feel free to merge after it passed.

vvcephei commented:
System tests passed: http://confluent-kafka-branch-builder-system-test-results.s3-us-west-2.amazonaws.com/2019-12-10--001.1576001003--vvcephei--smoke-test-fk--5332705/report.html

But, bummer, all the jenkins jobs just stopped running at some point. No errors or anything, just stopped about an 1:40 into the run. I'm guessing something bad happened to the containers the Jenkins agents run on.

Retest this, please.

vvcephei commented:
There was one unrelated test failure. Reported as https://issues.apache.org/jira/browse/KAFKA-6078
